### PR TITLE
Replace kube-dns with CoreDNS

### LINKED ISSUE #240: Replace kube-dns with CoreDNS
CoreDNS will be the new default DNS plugin for Kubernetes, hence, we should switch from kube-dns to CoreDNS as well.

### PR DESCRIPTION
**Pull request by zanetworker**
This pull request should replace kube-dns with CoreDNS as a core addon.

**What this PR does / why we need it**:

Currently Gardener uses `kube-dns` which is becoming deprecated for many reasons including having  3 containers in a pod doing the job of one + autoscaler, the containers are basically different tools (dnsmasq, kube-dns, and a side-car), which makes it difficult to maintain.  Therefore, with this PR, we replace the kube-dns component with [CoreDNS](https://coredns.io) which is more pluggable and which provides more features. 

**Which issue(s) this PR fixes**:
Fixes #240

**Special notes for your reviewer**:

The following tests were executed for this pull request: 
- [x] Testing DNS using a utils pod by DNSPinging a service.
- [x] Change the contents of a ConfigMap and verify that CoreDNS is still working.
- [x] Load-test the CoreDNS deployment (10 request / sec autoscaled to 10 nodes with 1000 replicas) to see if the horizontal pod auto-scaler works.
- [x]  Switching between master and feature branch (down-time happens only when reconciling, afterwards kube-dns gets replaced with CoreDNS smoothly without interrupting requests) 
- [x] verify all k8s versions are working, namely [1.11.2 1.10.7 1.9.10]

**Release note**:
```noteworthy user
Gardener does now replace `kube-dns` with [`CoreDNS`](https://coredns.io) in all Shoot clusters. `kube-dns` is becoming deprecated, and `CoreDNS` is the new default DNS plugin for Kubernetes (as it is more pluggable and provides more features). 
```

### CODE REVIEW

File: charts/shoot-core/charts/coredns/values.yaml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+          memory: 170Mi
+        requests:
+          cpu: 100m
+          memory: 70Mi
+
+configmap:
+  servers:
+  - zones:
+    - zone: .
+    port: 53
```
Comment (mvladev) [inline charts/shoot-core/charts/coredns/values.yaml]: Can we host it on unprivileged port, like `8053`? So we can drop `NET_BIND_SERVICE` capability.

**Review by mvladev — COMMENTED** [review]
You must also remove https://github.com/gardener/gardener/blob/master/charts/seed-monitoring/charts/prometheus/rules/kube-dnsmasq.rules.yml

File: charts/shoot-core/charts/coredns/values.yaml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+deployment:
+  spec:
+    containers:
+    # image:
+       # coredns: image-repository:image-tag
+      imagePullPolicy: IfNotPresent
+      resources:
+        limits:
+          cpu: 200m
+          memory: 170Mi
```
Comment (rfranzke) [inline charts/shoot-core/charts/coredns/values.yaml]: We have seen that this is not needed and much too high. Let's lower this to ~`80Mi`?

File: charts/shoot-core/charts/coredns/values.yaml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        requests:
+          cpu: 100m
+          memory: 70Mi
```
Comment (rfranzke) [inline charts/shoot-core/charts/coredns/values.yaml]: We have seen that this is not needed and much too high. Let's lower this to ~`15Mi`?

File: pkg/operation/hybridbotanist/addons.go [hunk 4]
```diff
@@ -80,7 +81,7 @@ func (b *HybridBotanist) generateCoreAddonsChart() (*chartrenderer.RenderedChart
 	if err != nil {
 		return nil, err
 	}
-	kubeDNS, err := b.Botanist.InjectImages(kubeDNSConfig, b.K8sShootClient.Version(), map[string]string{"kube-dns": "kube-dns", "kube-dns-dnsmasq": "kube-dns-dnsmasq", "kube-dns-sidecar": "kube-dns-sidecar", "kube-dns-autoscaler": "cluster-proportional-autoscaler"})
+	coreDNS, err := b.Botanist.InjectImages(coreDNSConfig, b.K8sSeedClient.Version(), map[string]string{"coredns": "coredns"})
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/addons.go]: Why have you changed from `K8sShootClient` to `K8sSeedClient`? The CoreDNS addon will be deployed into the Shoot, hence, we should keep using `K8sShootClient`.

**Review by rfranzke — CHANGES_REQUESTED** [review]
Thanks, looks very good overall.
Apart from the comments made inline, here a few TODOs:

1. Remove https://github.com/gardener/gardener/blob/master/pkg/features/gardener_features.go#L28-L32
2. Adapt https://github.com/gardener/gardener/blob/master/charts/seed-monitoring/charts/prometheus/rules/kube-dns.rules.yml approriately
3. Remove https://github.com/gardener/gardener/blob/master/charts/seed-monitoring/charts/prometheus/rules/kube-dnsmasq.rules.yml as pointed out by @mvladev .
4. Adapt https://github.com/gardener/gardener/blob/master/charts/shoot-core/values.yaml#L9-L15
5. Adapt comment in https://github.com/gardener/gardener/blob/master/charts/shoot-addons/charts/monocular/templates/api-deployment.yaml#L30
6. Adapt comment in https://github.com/gardener/gardener/blob/master/charts/shoot-addons/charts/nginx-ingress/values.yaml#L228
7. Remove https://github.com/gardener/gardener/blob/master/charts/shoot-addons/values.yaml#L6-L7 (not sure why it was there at all)
8. Adapt KubeDNS reference in https://raw.githubusercontent.com/gardener/gardener/master/docs/deployment/aks.md

Also, please rebase to the latest `master`.

File: charts/shoot-core/charts/coredns/values.yaml (hunk 1 above)
Comment (zanetworker) [reply charts/shoot-core/charts/coredns/values.yaml]: Hi @mvladev, 

Are you scrapping Kube-dns ? I am asking because I only found the job referenced in the rules, but nothing in the [config](https://github.com/gardener/gardener/blob/master/charts/seed-monitoring/charts/prometheus/templates/config.yaml)

Comment (mvladev) [reply charts/shoot-core/charts/coredns/values.yaml]: There is a template which uses the K8S service discovery in Prometheus (with the `prometheus.io/...` annotations) - https://github.com/gardener/gardener/blob/174ea1122aef8f39b37106fc2496b173be5fd04b/charts/seed-monitoring/charts/prometheus/templates/_helpers.tpl#L1-L29
It is being used in the Prometheus configuration:

https://github.com/gardener/gardener/blob/174ea1122aef8f39b37106fc2496b173be5fd04b/charts/seed-monitoring/charts/prometheus/templates/config.yaml#L232-L247

Comment (zanetworker) [reply charts/shoot-core/charts/coredns/values.yaml]: I see, thanks!

**Comment by zanetworker** [comment]
@rfranzke and @mvladev  thanks for reviewing. I have addressed your comments, please re-review to assert and let me know if further changes are required.

File: charts/images.yaml [hunk 5]
```diff
@@ -88,22 +88,10 @@ images:
...
-- name: kube-dns-sidecar
-  sourceRepository: github.com/kubernetes/dns
-  repository: k8s.gcr.io/k8s-dns-sidecar-amd64
-  tag: "1.14.10"
-- name: cluster-proportional-autoscaler
-  sourceRepository: github.com/kubernetes-incubator/cluster-proportional-autoscaler
-  repository: k8s.gcr.io/cluster-proportional-autoscaler-amd64
-  tag: "1.2.0"
+- name: coredns
+  sourceRepository: https://github.com/coredns/coredns
```
Comment (rfranzke) [inline charts/images.yaml]: Remove https://

File: charts/shoot-core/charts/coredns/templates/coredns-deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+      k8s-app: kube-dns
+  template:
+    metadata:
+      labels:
+        k8s-app: kube-dns
+      # we won't be using the checksum of the configmap since coredns provides the "reload" plugins that does the reload if config changes.
+    spec:
+      serviceAccountName: coredns
+      tolerations:
+      - key: node-role.kubernetes.io/master
```
Comment (rfranzke) [inline charts/shoot-core/charts/coredns/templates/coredns-deployment.yaml]: Not needed, we don't have master nodes.

File: charts/shoot-core/charts/coredns/values.yaml [hunk 7]
```diff
@@ -0,0 +1,77 @@
+# Default values for coredns.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+
```
Comment (rfranzke) [inline charts/shoot-core/charts/coredns/values.yaml]: Remove line 1-5.

File: charts/shoot-core/charts/coredns/values.yaml [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+# The values here match the ones on the upstream deployment repo (https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed). 
+# They may need to be updated in the future
+
+service:
+  type: "ClusterIP"
+  clusterDNS: 100.64.0.10 # This can changed if required, execute `kubectl get service --namespace kube-system kube-dns -o jsonpath="{.spec.clusterIP}` to get the clusterDNS IP
```
Comment (rfranzke) [inline charts/shoot-core/charts/coredns/values.yaml]: Remove the comment

File: charts/shoot-core/charts/coredns/values.yaml [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  #    contents: |
+  #      example.com.   IN SOA sns.dns.icann.com. noc.dns.icann.com. 2015082541 7200 3600 1209600 3600
+  #      example.com.   IN NS  b.iana-servers.net.
+  #      example.com.   IN NS  a.iana-servers.net.
+  #      example.com.   IN A   192.168.99.102
+  #      *.example.com. IN A   192.168.99.102
+
+horizontalPodAutoScaler:
+  spec:
+    maxReplicas: 3
```
Comment (rfranzke) [inline charts/shoot-core/charts/coredns/values.yaml]: Should we set it to 5 to be prepared for clusters having a high DNS load?

**Review by rfranzke — CHANGES_REQUESTED** [review]
Thanks!
Just found some minor cosmetic things. Apart from the comments made inline:

1. You still have to remove the kube-dns references here: https://github.com/gardener/gardener/blob/master/charts/shoot-core/values.yaml#L9-L15.
1. You still have to rename here: https://github.com/gardener/gardener/blob/master/charts/shoot-addons/charts/monocular/templates/api-deployment.yaml#L30
1. Please remove this whole section: https://github.com/gardener/gardener/blob/master/charts/shoot-addons/values.yaml#L1-L10
1. You still have to adapt https://github.com/gardener/gardener/blob/master/docs/deployment/aks.md ("./docs/deployment/aks.md:626:KubeDNS is running at ...")

File: pkg/openapi/openapi_generated.go [hunk 10]
```diff
@@ -13847,7 +13847,7 @@ func schema_k8sio_api_core_v1_ServiceSpec(ref common.ReferenceCallback) common.O
 					},
 					"externalName": {
 						SchemaProps: spec.SchemaProps{
-							Description: "externalName is the external reference that kubedns or equivalent will return as a CNAME record for this service. No proxying will be involved. Must be a valid RFC-1123 hostname (https://tools.ietf.org/html/rfc1123) and requires Type to be ExternalName.",
+							Description: "externalName is the external reference that CoreDNS or equivalent will return as a CNAME record for this service. No proxying will be involved. Must be a valid RFC-1123 hostname (https://tools.ietf.org/html/rfc1123) and requires Type to be ExternalName.",
```
Comment (rfranzke) [inline pkg/openapi/openapi_generated.go]: please revert this change, this file is autogenerated.

**Review by rfranzke — CHANGES_REQUESTED** [review]
Third try: Please remove this whole section: https://github.com/gardener/gardener/blob/master/charts/shoot-addons/values.yaml#L1-L10

Function-wise it looks good. Please check the above comment and the one made inline regarding the generated openapi file.

Also, please rebase on latest `master`. Other than these points I guess we can merge it afterwards.

**Review by rfranzke — APPROVED** [review]
/lgtm
